packages/remirror__extension-entity-reference/src/entity-reference-extension.ts

```diff
@@ -0,0 +1,264 @@
+import { decorateEntityReferences } from './utils/decorate-entity-references';
+import { getDisjoinedEntityReferencesFromNode } from './utils/disjoined-entity-references';
+import { joinDisjoinedEntityReferences } from './utils/joined-entity-references';
+
+/**
+ *  Required props to create entityReference marks decorations.
+ */
+interface StateProps {
+  extension: EntityReferenceExtension;
+  state: EditorState;
+}
+
+const getEntityReferencesFromPluginState = (props: StateProps) => {
+  const { extension, state } = props;
+  const { entityReferences } = extension.pluginKey.getState(state) as EntityReferencePluginState;
+  return entityReferences;
+};
+
+/**
+ * Creates a decoration set from the passed through state.
+ *
+ * @param props - see [[`DecorationProps`]]
+ */
+const createDecorationSet = (props: StateProps) => {
+  const { extension, state } = props;
+  const entityReferences = getEntityReferencesFromPluginState(props);
+  return DecorationSet.create(state.doc, extension.options.getStyle(entityReferences));
+};
+
+@extension<EntityReferenceOptions>({
+  defaultOptions: {
+    getStyle: decorateEntityReferences,
+    blockSeparator: undefined,
+    createId: uniqueId,
+  },
+})
+export class EntityReferenceExtension extends MarkExtension<EntityReferenceOptions> {
+  get name(): string {
+    return 'entity-reference' as const;
+  }
+
+  createMarkSpec(extra: ApplySchemaAttributes, override: MarkSpecOverride): MarkExtensionSpec {
+    const idAttr = `data-entityReference`;
+
+    return {
+      ...override,
+      excludes: '',
+      attrs: {
+        ...extra.defaults(),
+        id: { default: '' },
+      },
+      // Note that we don't implement `ParseDom`.
+      // The reasoning behind this is to disable pasting (parsing) the entityReference marks in the editor.
+      toDOM: (mark: Mark) => [
+        'span',
+        {
+          [idAttr]: mark.attrs.id,
+        },
+        0,
+      ],
+    };
+  }
+
+  /**
+   * Create the extension plugin for inserting decorations into the editor.
+   */
+  createPlugin(): CreateExtensionPlugin {
+    return {
+      state: {
+        init: (_: { [key: string]: any }, state: EditorState) => {
+          const entityReferences = this.getDisjoinedEntityReferences(state.doc);
+          return { entityReferences };
+        },
+        apply: (_tr: Transaction, _value: any, _oldState: EditorState, newState: EditorState) => {
+          const entityReferences = this.getDisjoinedEntityReferences(newState.doc);
+          return { entityReferences };
+        },
+      },
+      props: {
+        decorations: (state: EditorState) => {
+          return createDecorationSet({ state, extension: this });
+        },
+      },
+    };
+  }
+
+  @command()
+  addEntityReference(id?: string): CommandFunction {
+    return ({ state, tr, dispatch }) => {
+      const { from, to } = state.selection;
+      const newId = id ?? (this.options.createId ? this.options.createId() : uniqueId());
```

Comment:
`createId` already defaults to `uniqueId`

```suggestion
      const newId = id ?? this.options.createId();
```

Reply:
Indeed, this will be obsolete when we remove `createId`